OpenGL linking error on Windows

The library isn't linked against OpenGL which causes this error:
```
ld.exe: scratchcpp-render/src/libscratchcpp-render.a(renderedtarget.cpp.obj):renderedtarget:(.text+0x276c): undefined reference to `__imp_glReadPixels'
```
However, Qt is linked against OpenGL, so QOpenGLFunctions should be used instead of directly calling the functions.

--- a/src/penlayer.cpp
+++ b/src/penlayer.cpp
@@ -13,6 +13,8 @@ PenLayer::PenLayer(QNanoQuickItem *parent) :
 {
     m_fboFormat.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);
     m_fboFormat.setSamples(m_antialiasingEnabled ? 4 : 0);
+
+    m_glF.initializeOpenGLFunctions();
 }
 
 PenLayer::~PenLayer()
@@ -65,10 +67,10 @@ void scratchcpprender::PenLayer::clear()
         return;
 
     m_fbo->bind();
-    glDisable(GL_SCISSOR_TEST);
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
-    glEnable(GL_SCISSOR_TEST);
+    m_glF.glDisable(GL_SCISSOR_TEST);
+    m_glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    m_glF.glClear(GL_COLOR_BUFFER_BIT);
+    m_glF.glEnable(GL_SCISSOR_TEST);
     m_fbo->release();
 
     update();

--- a/src/penlayer.h
+++ b/src/penlayer.h
@@ -5,6 +5,7 @@
 #include <ipenlayer.h>
 #include <QOpenGLFramebufferObject>
 #include <QOpenGLPaintDevice>
+#include <QOpenGLFunctions>
 #include <QPainter>
 #include <scratchcpp/iengine.h>
 
@@ -49,6 +50,7 @@ class PenLayer : public IPenLayer
         std::unique_ptr<QOpenGLFramebufferObject> m_fbo;
         std::unique_ptr<QOpenGLPaintDevice> m_paintDevice;
         QOpenGLFramebufferObjectFormat m_fboFormat;
+        QOpenGLFunctions m_glF;
 };
 
 } // namespace scratchcpprender

--- a/src/renderedtarget.cpp
+++ b/src/renderedtarget.cpp
@@ -513,6 +513,12 @@ void RenderedTarget::clearGraphicEffects()
 void RenderedTarget::updateHullPoints(QOpenGLFramebufferObject *fbo)
 {
     Q_ASSERT(fbo);
+
+    if (!m_glF) {
+        m_glF = std::make_unique<QOpenGLFunctions>();
+        m_glF->initializeOpenGLFunctions();
+    }
+
     int width = fbo->width();
     int height = fbo->height();
     m_hullPoints.clear();
@@ -521,7 +527,7 @@ void RenderedTarget::updateHullPoints(QOpenGLFramebufferObject *fbo)
     // Read pixels from framebuffer
     size_t size = width * height * 4;
     GLubyte *pixelData = new GLubyte[size];
-    glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pixelData);
+    m_glF->glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pixelData);
 
     // Flip vertically
     int rowSize = width * 4;

--- a/src/renderedtarget.h
+++ b/src/renderedtarget.h
@@ -125,6 +125,7 @@ class RenderedTarget : public IRenderedTarget
         Skin *m_skin = nullptr;
         Texture m_texture;
         Texture m_oldTexture;
+        std::unique_ptr<QOpenGLFunctions> m_glF;
         std::unordered_map<ShaderManager::Effect, double> m_graphicEffects;
         double m_size = 1;
         double m_x = 0;

--- a/src/skin.cpp
+++ b/src/skin.cpp
@@ -18,6 +18,9 @@ Texture Skin::createAndPaintTexture(int width, int height, bool multisampled)
     if (!context || !context->isValid() || (width <= 0 || height <= 0))
         return Texture();
 
+    QOpenGLFunctions glF(context);
+    glF.initializeOpenGLFunctions();
+
     // Create offscreen surface
     QOffscreenSurface surface;
     surface.setFormat(context->format());
@@ -48,8 +51,8 @@ Texture Skin::createAndPaintTexture(int width, int height, bool multisampled)
     QPainter painter(&device);
     painter.beginNativePainting();
     painter.setRenderHint(QPainter::Antialiasing, false);
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
+    glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glF.glClear(GL_COLOR_BUFFER_BIT);
 
     // Call the skin-specific paint method
     paint(&painter);

--- a/src/texture.cpp
+++ b/src/texture.cpp
@@ -114,8 +114,12 @@ QImage Texture::toImage() const
 
 void Texture::release()
 {
-    if (m_isValid) {
-        glDeleteTextures(1, &m_handle);
+    QOpenGLContext *context = QOpenGLContext::currentContext();
+
+    if (m_isValid && context) {
+        QOpenGLExtraFunctions glF(context);
+        glF.initializeOpenGLFunctions();
+        glF.glDeleteTextures(1, &m_handle);
         m_isValid = false;
     }
 }

--- a/test/targetpainter/targetpainter_test.cpp
+++ b/test/targetpainter/targetpainter_test.cpp
@@ -37,6 +37,9 @@ TEST_F(TargetPainterTest, Paint)
     QOpenGLFramebufferObjectFormat format;
     format.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);
 
+    QOpenGLExtraFunctions glF(&m_context);
+    glF.initializeOpenGLFunctions();
+
     // Begin painting reference
     QNanoPainter refPainter;
     QOpenGLFramebufferObject refFbo(40, 60, format);
@@ -87,8 +90,8 @@ TEST_F(TargetPainterTest, Paint)
     ASSERT_EQ(fbo.toImage(), refFbo.toImage());
 
     // Paint with color effects
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
+    glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glF.glClear(GL_COLOR_BUFFER_BIT);
     effects[ShaderManager::Effect::Color] = 46;
     effects[ShaderManager::Effect::Brightness] = 20;
     effects[ShaderManager::Effect::Ghost] = 84;

--- a/test/textbubblepainter/textbubblepainter_test.cpp
+++ b/test/textbubblepainter/textbubblepainter_test.cpp
@@ -33,6 +33,9 @@ TEST_F(TextBubblePainterTest, PaintSayBubble)
     QOffscreenSurface surface;
     createContextAndSurface(&context, &surface);
 
+    QOpenGLExtraFunctions glF(&context);
+    glF.initializeOpenGLFunctions();
+
     QOpenGLFramebufferObjectFormat format;
     format.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);
 
@@ -40,8 +43,8 @@ TEST_F(TextBubblePainterTest, PaintSayBubble)
     QNanoPainter painter;
     QOpenGLFramebufferObject fbo(425, 250, format);
     fbo.bind();
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+    glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glF.glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
     painter.beginFrame(fbo.width(), fbo.height());
     painter.setAntialias(0);
 
@@ -77,6 +80,9 @@ TEST_F(TextBubblePainterTest, PaintThinkBubble)
     QOffscreenSurface surface;
     createContextAndSurface(&context, &surface);
 
+    QOpenGLExtraFunctions glF(&context);
+    glF.initializeOpenGLFunctions();
+
     QOpenGLFramebufferObjectFormat format;
     format.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);
 
@@ -84,8 +90,8 @@ TEST_F(TextBubblePainterTest, PaintThinkBubble)
     QNanoPainter painter;
     QOpenGLFramebufferObject fbo(156, 117, format);
     fbo.bind();
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+    glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glF.glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
     painter.beginFrame(fbo.width(), fbo.height());
     painter.setAntialias(0);
 

--- a/test/texture/texture_test.cpp
+++ b/test/texture/texture_test.cpp
@@ -49,6 +49,9 @@ TEST(TextureTest, ToImage)
     Q_ASSERT(surface.isValid());
     context.makeCurrent(&surface);
 
+    QOpenGLExtraFunctions glF(&context);
+    glF.initializeOpenGLFunctions();
+
     QOpenGLFramebufferObjectFormat format;
     format.setAttachment(QOpenGLFramebufferObject::CombinedDepthStencil);
 
@@ -59,8 +62,8 @@ TEST(TextureTest, ToImage)
     QPainter painter(&device);
     painter.beginNativePainting();
     painter.setRenderHint(QPainter::Antialiasing, false);
-    glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
+    glF.glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+    glF.glClear(GL_COLOR_BUFFER_BIT);
     painter.drawEllipse(0, 0, fbo.width(), fbo.height());
     painter.endNativePainting();
     painter.end();
@@ -86,15 +89,18 @@ TEST(TextureTest, Release)
     Q_ASSERT(surface.isValid());
     context.makeCurrent(&surface);
 
+    QOpenGLExtraFunctions glF(&context);
+    glF.initializeOpenGLFunctions();
+
     QOpenGLFramebufferObject fbo(1, 1);
     GLuint handle = fbo.takeTexture();
-    ASSERT_TRUE(glIsTexture(handle));
+    ASSERT_TRUE(glF.glIsTexture(handle));
 
     Texture tex(handle, fbo.width(), fbo.height());
-    ASSERT_TRUE(glIsTexture(handle));
+    ASSERT_TRUE(glF.glIsTexture(handle));
 
     tex.release();
-    ASSERT_FALSE(glIsTexture(handle));
+    ASSERT_FALSE(glF.glIsTexture(handle));
     ASSERT_FALSE(tex.isValid());
 
     context.doneCurrent();

